src/compiler/checker.ts

```diff
@@ -5106,6 +5161,53 @@ namespace ts {
             }
         }
 
+        /**
+         * Gets the minimum number of type arguments needed to satisfy all non-optional type
+         * parameters.
+         */
+        function getMinTypeArgumentCount(typeParameters: TypeParameter[] | undefined): number {
+            let minTypeArgumentCount = 0;
+            if (typeParameters) {
+                for (let i = 0; i < typeParameters.length; i++) {
+                    if (!getDefaultFromTypeParameter(typeParameters[i])) {
+                        minTypeArgumentCount = i + 1;
+                    }
+                }
+            }
+            return minTypeArgumentCount;
+        }
+
+        function fillMissingTypeArguments(typeArguments: Type[] | undefined, typeParameters: TypeParameter[] | undefined, minTypeArgumentCount: number) {
+            const numTypeArguments = typeArguments ? typeArguments.length : 0;
+            const numTypeParameters = typeParameters ? typeParameters.length : 0;
+            if (numTypeArguments >= minTypeArgumentCount && numTypeArguments <= numTypeParameters) {
+                if (numTypeParameters) {
+                    if (!typeArguments) {
+                        typeArguments = [];
+                    }
+                    const mapper: TypeMapper = t => {
+                        for (let i = 0; i < numTypeParameters; i++) {

```

Comment:
You technically don't need to run through the initial type arguments that were already supplied, do you? Could you just start at `numTypeArguments` as you did below?